pkg/apis/configuration/v1/kongclusterplugin_types.go

```diff
@@ -64,6 +64,13 @@ type KongClusterPlugin struct {
 	// Only one of `config` or `configFrom` may be used in a KongClusterPlugin, not both at once.
 	ConfigFrom *NamespacedConfigSource `json:"configFrom,omitempty"`
 
+	// ConfigPatches represents JSON patches to the configuration of the plugin.
+	// Each item means a JSON patch to add something in the configuration,
+	// where path is specified in `path` and value is in `valueFrom` referencing
+	// a key in a secret.
+	// Could only be specified when Config specified.
```

Comment:
Please add a CEL rule that would prevent `ConfigPatches` from being set when there's no `Config` set.

Reply:
Do we need this restriction in general? It's not clear to me that you should. Patches can be applied to an empty object, and may be easier to write that way depending on the plugin schema.

If, for example, you want to add a two sensitive header using https://docs.konghq.com/hub/kong-inc/request-transformer/configuration/#config-add, it'd be reasonable to create a patch for `/add` with the complete headers value, rather than needing to provide an empty `headers` array and individual values:

```
{"otheradd":{"headers":[]}}
```

```
[
    { "op": "add", "path": "/otheradd/headers/0", "value": "foo:red" },
    { "op": "add", "path": "/otheradd/headers/1", "value": "bar:blue" },
    { "op": "add", "path": "/add", "value": {"headers":["ha:hello", "hb:world"]} }
]
```

Which you use is mostly a matter of preference, though there are probably some cases where the syntax is easier one way or the other.

Not exactly sure how this interacts with the `omitempty` directive and whether assuming `{}` on empty creates issues with the existing `configFrom` checks. If we need it for that, can you manually specify the empty object, or does that get eaten also?

Reply:
@rainest Yeah, I think you're making a good point about the use case with multiple fields that are coming from different sources (i.e. different Secrets) and there's no plain text config at all. I agree we should account for that. It renders the mentioned constraint of `ConfigPatches` being allowed only when `Config` is set as not necessary.

We should probably have a CEL rule ensuring `ConfigPatches` is not set when `ConfigFrom` is set though OR handle it accordingly in code (as I can see from the code it wouldn't work right now as `ConfigFrom` Secret would overwrite what comes from `ConfigPatches`). I'd lean towards making use cases limited so the first option seems more appealing to me.

Reply:
Changed the CEL rules to limit it not to be used with `configFrom`. \

But we still could not support "empty `config` and `configPatches` to set the whole config", we could only achieve that by specifying `config` to `{}` (empty object) and use `path` to `""` (empty string).

Reply:
How about such an approach https://github.com/Kong/kubernetes-ingress-controller/pull/5167? It makes it possible to use empty `config` with `configPatches`.

Reply:
Updated this: we cannot achieve it by `add` operation (possibly issues in the json-patch package). Supported by using `replace` on root to set whole config.

Reply:
No path is useful as a complete replacement for `ConfigFrom` if you do want to store _everything_ in a Secret (or were maybe just already using `ConfigFrom` and want to do a lazy migration).

If that requires `replace` though, we can't do that without making the configuration more complex, so requiring individual implicit `add` directives is acceptable.